### PR TITLE
Fix inflight orders log

### PR DESCRIPTION
**Pull request by sunce86**
Two changes in this PR:
1. Prints more detailed logs around inflight orders, and not just when the orders from auction are reduced but always.
2. Set's the higher transaction isolation level for `solvable_orders` because of the issue explained [here](https://github.com/cowprotocol/services/pull/1819#issuecomment-1696146145). Basically, two subsequent `SELECT` queries are not guaranteed to see the same database state with the default transaction isolation level.

### Release notes

Need to set the log level for inflight_orders.rs to trace.

### CODE REVIEW

**Review by fleupold — COMMENTED** [review]
I'm a bit worried that this log will be huge (we have around 1k orders in the auction). I also don't think it's helping us for the issue at hand, as the important question is why did the autopilot not filter this order as filled.

**Comment by sunce86** [comment]
> I'm a bit worried that this log will be huge
The log is printing inflight orders (not all orders in the auction) which is usually around several.

> I also don't think it's helping us for the issue at hand, as the important question is why did the autopilot not filter this order as filled.
My primary goal here was to have a place in the code where the `latest_settlement_block` will be printed since it's not printed anywhere but could tell us (and future us) if the event indexing happened before solvable orders cache update or not.

**Review by fleupold — APPROVED** [review]
Oops I missed that this just removed the conditionality. Looks good then, although it will probably just tell us that inflight order_uids didn't contain the order we were looking for (it will also get mutated during update_and_filter so may remove orders that are stored at an old block number. Maybe we should also add a log where the cache is populated to know what order is associated with what block number?

**Comment by fleupold** [comment]
> My primary goal here was to have a place in the code where the latest_settlement_block is printed

It's part of the auction that we store in /api/v1/solver_competition/{auction_id}

**Comment by sunce86** [comment]
> > My primary goal here was to have a place in the code where the latest_settlement_block is printed
> 
> It's part of the auction that we store in /api/v1/solver_competition/{auction_id}

If you meant `auctionStartBlock` that's not equal to `auction.latest_settlement_block`?

**Comment by sunce86** [comment]
> Maybe we should also add a log where the cache is populated to know what order is associated with what block number

If you meant cache for inflight orders, for that I can assume how it is populated, with high confidence, since the block number is taken from the `tx_receipt.block_number` which we can trust.
Also, I know orders are inserted to inflight for the auction `7270437` because of [this log](https://production-6de61f.kb.eu-central-1.aws.cloud.es.io/app/r/s/5CPyx) which is executed in the same code block as insertion to inflight.

**Comment by fleupold** [comment]
Getting orders and fetching latest_settlement_block happens in two different db calls so you don't have read consistency between them:
https://github.com/cowprotocol/services/blob/f9cfba2549bcadd73467cc00e2b229eab9fa133a/crates/autopilot/src/database/auction.rs#L81-L93

Maybe the right fix for now would be to return an error in the update task if `db_solvable_orders.latest_settlement_block` > `block`

**Comment by fleupold** [comment]
Actually the default transaction isolation level may allow for this type of race condition. From https://www.postgresql.org/docs/current/transaction-iso.html

> Read Committed is the default isolation level in PostgreSQL. When a transaction uses this isolation level, a SELECT query (without a FOR UPDATE/SHARE clause) sees only data committed before the query began; it never sees either uncommitted data or changes committed during query execution by concurrent transactions. In effect, a SELECT query sees a snapshot of the database as of the instant the query begins to run. However, SELECT does see the effects of previous updates executed within its own transaction, even though they are not yet committed. **Also note that two successive SELECT commands can see different data, even though they are within a single transaction, if other transactions commit changes after the first SELECT starts and before the second SELECT starts.**

My hunch is that this is what happened here...

**Comment by sunce86** [comment]
> Actually the default transaction isolation level may allow for this type of race condition. From https://www.postgresql.org/docs/current/transaction-iso.html
> 
> > Read Committed is the default isolation level in PostgreSQL. When a transaction uses this isolation level, a SELECT query (without a FOR UPDATE/SHARE clause) sees only data committed before the query began; it never sees either uncommitted data or changes committed during query execution by concurrent transactions. In effect, a SELECT query sees a snapshot of the database as of the instant the query begins to run. However, SELECT does see the effects of previous updates executed within its own transaction, even though they are not yet committed. **Also note that two successive SELECT commands can see different data, even though they are within a single transaction, if other transactions commit changes after the first SELECT starts and before the second SELECT starts.**
> 
> My hunch is that this is what happened here...

Oh boy... this seems like it could definitely a problem which I was not aware of until now. Whenever I used the transaction I assumed the consistent view from the beginning of the transaction.

Anyway, I addressed this https://github.com/cowprotocol/services/pull/1819/commits/e09bf81a1c6b27204fd02220d15c14c56d5a1a6e
But it is probably worth discussing if we want to change the global configuration to avoid future similar cases.

**Comment by fleupold** [comment]
I think we should. Interested to hear @MartinquaXD and @nlordell's thoughts. We could add a flyway migration that does something like [this](https://stackoverflow.com/questions/62649971/how-to-change-transaction-isolation-level-globally)

**Comment by MartinquaXD** [comment]
Interesting read. 🕵️ 
> We could add a flyway migration that does something like [this](https://stackoverflow.com/questions/62649971/how-to-change-transaction-isolation-level-globally)

My initial gut feeling is that changing such a setting system wide might produce unexpected consequences. My initial worry is that increasing the guarantees of the transaction isolation level might result in worse performance but according to SO it should (:tm:) be okay. To reduce the chances for unforeseen problems to a minimum I would suggest to try the [repeatable read](https://www.postgresql.org/docs/current/transaction-iso.html#XACT-REPEATABLE-READ) level which guarantees what my expectation of a transaction were while not imposing any additional (potentially costly) guarantees.
One thing to note on that matter is this:
> Applications using this level must be prepared to retry transactions due to serialization failures.

We have surprisingly few transactions (outside of tests) and I think most of them should not be affected by that serialization error mention in the docs. So overall the hassle involved in updating the isolation level should (:tm:) not be so much.

File: crates/solver/src/in_flight_orders.rs [hunk 1]
```diff
@@ -89,6 +93,19 @@ impl InFlightOrders {
...
             }
         });
+
+        tracing::debug!(
+            auction_block = %auction.block,
+            latest_settlement_block = %auction.latest_settlement_block,
+            inflight_before = %inflight_before.len(),
+            inflight_before = ?inflight_before,
+            inflight_after = %in_flight.len(),
+            inflight_after = ?in_flight,
```
Comment (MartinquaXD) [inline crates/solver/src/in_flight_orders.rs:103]: The keys `inflight_before` and `inflight_after` are duplicated with different values.

File: crates/solver/src/in_flight_orders.rs [hunk 2]
```diff
@@ -89,6 +93,19 @@ impl InFlightOrders {
                 u256_to_big_uint(&order.data.buy_amount) > order.metadata.executed_buy_amount
             }
         });
+
+        tracing::debug!(
+            auction_block = %auction.block,
+            latest_settlement_block = %auction.latest_settlement_block,
+            inflight_before = %inflight_before.len(),
+            inflight_before = ?inflight_before,
```
Comment (MartinquaXD) [inline crates/solver/src/in_flight_orders.rs:101]: nit: Listing the order UIDs might get long. I would suggest to move all the short values (blocks and lengths) to the front so the logs shows as much information as possible on a first glance without having to expand the log in kibana.

**Review by MartinquaXD — APPROVED** [review]
LGTM
just nits

File: crates/solver/src/in_flight_orders.rs [hunk 3]
```diff
@@ -89,6 +93,19 @@ impl InFlightOrders {
                 u256_to_big_uint(&order.data.buy_amount) > order.metadata.executed_buy_amount
             }
         });
+
+        tracing::debug!(
```
Comment (nlordell) [inline crates/solver/src/in_flight_orders.rs:97]: Personally, I would demote this to a `trace` log. We can enable `trace` log level for this module for now.

**Review by nlordell — APPROVED** [review]
LGTM - I would also keep this at a transaction and not global level as @MartinquaXD suggested.

**Comment by nlordell** [comment]
nit: Can you update the PR description and title to reflect the changes to the database read semantics (its not just adding logs but changing db transaction behaviour).

File: crates/solver/src/in_flight_orders.rs (hunk 1 above)
Comment (sunce86) [reply crates/solver/src/in_flight_orders.rs:103]: This was intentional, I am not aware of any issues this would cause?

Comment (MartinquaXD) [reply crates/solver/src/in_flight_orders.rs:103]: Interesting. Didn't know that you can set the same key twice without overriding the value. 👌

**Comment by fleupold** [comment]
I updated the name of the PR now, but it's already been merged with the old name. We have to remember now that we did this if we browse the release title next week (no-one would expect such a meaty change just from the title)

**Pull request merged**